torchft/process_group.py

```diff
@@ -589,15 +589,21 @@ def __init__(self, pg: ProcessGroup, work: Work, timeout: timedelta) -> None:
         self._timeout = timeout
 
     def wait(self, timeout: Optional[timedelta] = None) -> bool:
-        with self._stream_timeout(self._pg, self._timeout):
-            # not async, just return
-            if self._work is None:
-                return True
-
+        async_timeout = timeout or self._timeout
+        with self._stream_timeout(self._pg, async_timeout):
+            # In newer versions of PyTorch work may not exist if the call was
+            # not async. In these cases we can just schedule the stream timeout
+            # and return.
+            if self._work is not None:
+                if not self._work.wait():
+                    return False
+
+            # Always use cuda stream for timeout to avoid ProcessGroupNCCL
+            # watchdog firing and crashing the process.
             if timeout is not None:
-                return self._work.wait(timeout)
+                torch.cuda.synchronize()
```

Comment:
For my understanding, we are calling `torch.cuda.synchronize()` here so that it will hang and timeout in   `_stream_timeout` which uses `stream_timeout` (https://github.com/pytorch/torchft/blob/main/torchft/process_group.py#L627) to handle the error, is that right? What happens in the case that timeout is `None`?

Reply:
In the case that timeout is not None we will instead synchronize in `should_commit`. The stream_timeout logic will still apply and call `abort` if the stream hangs so we'll detect the failure and report it in should_commit.